File: graphql/udemy1-graphql17/GraphQL-JWT-authentication/src/main/java/com/accounts/interceptor/RequestHeaderInterceptor.java
package com.accounts.interceptor;

import com.accounts.config.JwtUtils;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.graphql.server.WebGraphQlInterceptor;
import org.springframework.graphql.server.WebGraphQlRequest;
import org.springframework.graphql.server.WebGraphQlResponse;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

@Component
@Slf4j
public class RequestHeaderInterceptor implements WebGraphQlInterceptor {

    @Autowired
    JwtUtils jwtUtils;

    @Override
    public Mono<WebGraphQlResponse> intercept(WebGraphQlRequest request, Chain chain) {
        if (request.getDocument().contains("query") && request.getDocument().contains("login")) {
            return chain.next(request);
        }else{
            String bearerToken = request.getHeaders().getFirst("Authorization");

            log.debug("Authorization Header: {}", bearerToken);
            if (bearerToken != null && bearerToken.startsWith("Bearer ")) {
                if (jwtUtils.validateJwtToken(bearerToken)) {
                    return chain.next(request);
                }else{
                    return Mono.error(new RuntimeException("Invalid credentials"));
                }
            }else{
                return Mono.error(new RuntimeException("Invalid credentials"));
            }
        }
    }
}

File: graphql/graphqlex/ex1/src/main/kotlin/com/yscorp/ex1/interceptor/RequestHeaderInterceptor.kt
package com.yscorp.ex1.interceptor

import com.yscorp.ex1.config.JwtUtils
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.graphql.server.WebGraphQlInterceptor
import org.springframework.graphql.server.WebGraphQlRequest
import org.springframework.graphql.server.WebGraphQlResponse
import org.springframework.stereotype.Component
import reactor.core.publisher.Mono

@Component
class RequestHeaderInterceptor(
    var jwtUtils: JwtUtils,
) : WebGraphQlInterceptor {


    private val log = KotlinLogging.logger { }

    override fun intercept(request: WebGraphQlRequest, chain: WebGraphQlInterceptor.Chain): Mono<WebGraphQlResponse> {
        if (request.document.contains("query") && request.document.contains("login")) {
            return chain.next(request)
        } else {
            val bearerToken = request.headers.getFirst("Authorization")

            log.debug("Authorization Header: {}", bearerToken)

            return if (bearerToken != null && bearerToken.startsWith("Bearer ")) {
                if (jwtUtils.validateJwtToken(bearerToken)) {
                    chain.next(request)
                } else {
                    Mono.error(RuntimeException("Invalid credentials"))
                }
            } else {
                Mono.error(RuntimeException("Invalid credentials"))
            }
        }
    }
}